kms/tpmkms/tpmkms.go

```diff
@@ -0,0 +1,523 @@
+	ekKeyID, err := generateKeyID(ekPublic)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting EK public key ID: %w", err)
+	}
+	ekKeyURL := ekURL(ekKeyID)
+
+	// check if the derived EK URI fingerprint representation matches the provided
+	// permanent identifier value. The current implementation requires the EK URI to
+	// be used as the AK identity, so an error is returned if there's no match. This
+	// could be changed in the future, so that another attestation flow takes place,
+	// instead, for example.
+	if k.permanentIdentifier != "" && !strings.EqualFold(ekKeyURL.String(), k.permanentIdentifier) {
+		return nil, fmt.Errorf("the provided permanent identifier %q does not match the EK URL %q", k.permanentIdentifier, ekKeyURL.String())
+	}
+
+	// check if a (valid) AK certificate (chain) is available. Perform attestation flow
+	// otherwise. If an AK certificate is available, but not considered valid, e.g. due
+	// to it not having the right identity, a new attestation flow will be performed and
+	// the old certificate (chain) will be overwritten with the result of that flow.
+	akChain := ak.CertificateChain()
+	if len(akChain) == 0 || !hasValidIdentity(ak, ekKeyURL) {
+		var ac apiv1.AttestationClient
+		if req.AttestationClient != nil {
+			// TODO(hs): check if it makes sense to have this; it doesn't capture all
+			// behavior of the built-in attestorClient, but at least it does provide
+			// a basic extension point for other ways of performing attestation that
+			// might be useful for testing or attestation flows against other systems.
+			// For it to be truly useful, the logic for determining the AK identity
+			// would have to be updated too, though.
+			ac = req.AttestationClient
+		} else {
+			ac, err = k.newAttestorClient(ek, ak)
+			if err != nil {
+				return nil, fmt.Errorf("failed creating attestor client: %w", err)
+			}
+		}
+		// perform the attestation flow with a (remote) attestation CA
+		if akChain, err = ac.Attest(ctx); err != nil {
+			return nil, fmt.Errorf("failed performing AK attestation: %w", err)
+		}
+		// store the result with the AK, so that it can be reused for future
+		// attestations.
+		if err := ak.SetCertificateChain(ctx, akChain); err != nil {
+			return nil, fmt.Errorf("failed storing AK certificate chain: %w", err)
+		}
+	}
+
+	// when a new certificate was issued for the AK, it is possible the
+	// certificate that was issued doesn't include the expected and/or required
+	// identity, so this is checked before continuing.
+	if !hasValidIdentity(ak, ekKeyURL) {
+		return nil, fmt.Errorf("AK certificate (chain) not valid for EK %q", ekKeyURL)
+	}
+
+	// prepare the response to return
+	akCert := akChain[0]
+	permanentIdentifier := ekKeyURL.String() // TODO(hs): should always match the valid value of the AK identity (for now)
+	resp := &apiv1.CreateAttestationResponse{
+		Certificate:         akCert,
+		CertificateChain:    akChain, // TODO(hs): should this include the leaf or not?
```

Comment:
Yubikey, which is the only one implementing this does not include the leaf.

Reply:
In other places where we refer to a chain, we include the leaf as the first certificate. I think it's best to follow that? But if we do, `Certificate` is not required. In other places where attestation is done using certificates, I think there's usually a chain, which can be just the leaf. E.g. WebAuthn uses the `x5c` COSE claim for that. IMO defaulting to (just) the chain thus makes sense.

Reply:
I've updated the `YubiKey` KMS to include the leaf as the first cert in the chain. This is not backwards compatible, but it's probably OK to do so.